### ISSUE TITLE
ESM build : reassignment of import and require

### ISSUE THREAD

**Issue by camille-hdl** [alex3165/react-mapbox-gl#795]
Hello,

i'm trying to use react-mapbox-gl 4 with Rollup but I'm having problems :

1. The problem [discussed here](https://github.com/rollup/rollup/issues/3187) - [how to reproduce](https://github.com/camille-hdl/rollup-starter-lib)

In short:

In `node_modules/react-mapbox-gl/lib-esm/map.js`, there is

```js
import * as MapboxGl from 'mapbox-gl';
/* ... */
MapboxGl.accessToken = accessToken;
```
which throws `Illegal reassignment to import 'MapboxGl'`. 

Since mapbox-gl has a default export, changing the first line to `import MapboxGl from 'mapbox-gl';` fixes this problem.

2. A few lines later there is `var isEqual = require('deep-equal');`, which is causing problems since  there is mixed ESM and CommonJS syntax in the same file, and this is the "module" entrypoint in package.json.

Of course I may be missing something here. If so, if someone has any experience getting react-mapbox-gl 4 working with Rollup, I'd appreciate any help.

For what it's worth, version 3.x works fine (but there is no esm build).

**Comment by mklopets** [comment]
does the 4.x commonjs version work for you?

**Comment by gmaclennan** [comment]
Hi I ran into what I think is a related issue when bundling with Webpack - the setter was not being called so `MapboxGl.accessToken` was set to my access token, but [this setter](https://github.com/mapbox/mapbox-gl-js/blob/d69e24420d9ef4e69315d5dc590ee7e047706aa2/src/index.js#L59) was not called, so `config.ACCESS_TOKEN` was never set. This only started becoming a problem in the past month, so I am assuming that Webpack made a change to be more spec compliant and stop this working?

TL;DR I think the way this is written in ESM should not work, but it works in many bundlers because they are a little more lenient. I think the change @camille-hdl suggests should work.

**Comment by camille-hdl** [comment]
@mklopets I tried manually importing from the cjs build, but ran into other issues (I don't remember exactly but it was related to the mixed `import` / `require` syntax I described in the OP iirc)

**Comment by gmaclennan** [comment]
ah ok I see the code is written in TypeScript, so you need this `*` import to get the types. I'm not familiar enough with TypeScript to know the workaround for this.

**Comment by nrakochy** [comment]
Not sure if I was looking at the same issue, but you might try installing the `mapbox-gl` types directly
```js
npm install -D @types/mapbox-gl 
```

**Comment by sirgalleto** [comment]
@camille-hdl did you manage to fix this?

**Comment by camille-hdl** [comment]
@sirgalleto I haven't found a workaround, that said I haven't tried to update it in a while (I stuck to 3.x for now), so I don't know if it's fixed now.

**Comment by sirgalleto** [comment]
@camille-hdl I was trying to use the latest version, but is still broken, now using 3.x as well. 

I'm wondering what options we have for TypeScript, I have not too much experience using TS so I'm not sure about how to change the `import * from MapgoxGL`. I'll investigate a little bit and try to come with a solution.

**Comment by camille-hdl** [comment]
In case anyone stumbles upon this, I managed to get react-mapbox-gl 5 to work with React 19 and Rollup 5 with the following changes:

First, apply a patch using [patch-package](https://www.npmjs.com/package/patch-package):

patches/react-mapbox-gl+5.1.1.patch
```patch
diff --git a/node_modules/react-mapbox-gl/lib-esm/geojson-layer.js b/node_modules/react-mapbox-gl/lib-esm/geojson-layer.js
index 2d78d03..3c38487 100644
--- a/node_modules/react-mapbox-gl/lib-esm/geojson-layer.js
+++ b/node_modules/react-mapbox-gl/lib-esm/geojson-layer.js
@@ -23,7 +23,7 @@ var __assign = (this && this.__assign) || function () {
     return __assign.apply(this, arguments);
 };
 import * as React from 'react';
-var isEqual = require('deep-equal');
+import isEqual from 'deep-equal';
 import diff from './util/diff';
 import { generateID } from './util/uid';
 import { withMap } from './context';
diff --git a/node_modules/react-mapbox-gl/lib-esm/layer.js b/node_modules/react-mapbox-gl/lib-esm/layer.js
index 5a6bff6..717af01 100644
--- a/node_modules/react-mapbox-gl/lib-esm/layer.js
+++ b/node_modules/react-mapbox-gl/lib-esm/layer.js
@@ -23,7 +23,7 @@ var __assign = (this && this.__assign) || function () {
     return __assign.apply(this, arguments);
 };
 import * as React from 'react';
-var isEqual = require('deep-equal');
+import isEqual from 'deep-equal';
 import diff from './util/diff';
 var eventToHandler = {
     touchstart: 'onTouchStart',
diff --git a/node_modules/react-mapbox-gl/lib-esm/map.js b/node_modules/react-mapbox-gl/lib-esm/map.js
index a475a36..adcb4ec 100644
--- a/node_modules/react-mapbox-gl/lib-esm/map.js
+++ b/node_modules/react-mapbox-gl/lib-esm/map.js
@@ -27,7 +27,8 @@ import * as React from 'react';
 import { listenEvents, events, updateEvents } from './map-events';
 import { MapContext } from './context';
 import { createPortal } from 'react-dom';
-var isEqual = require('deep-equal');
+// var isEqual = require('deep-equal');
+import isEqual from 'deep-equal';
 var defaultZoom = [11];
 var defaultMovingMethod = 'flyTo';
 var defaultCenter = [-0.2416815, 51.5285582];
```

Then, using @rollup/plugin-replace in my Rollup config:

```js
replace({
            "require.main": () => {
                console.log("WARNING: replace require.main by import.meta");
                return "import.meta"; // you can also just return `false` or whatever.
            },
}),
```

That being said, migrating away from this package is now a priority as this is clearly not a long term solution. I did this ugly thing in a rush because version 3 uses the old Context API which breaks in React 19.

Good luck
